tools/test/resources/resource_test.py

```diff
@@ -0,0 +1,136 @@
+        Verify that the ordering of Target info addition and directory addition
+        does not matter, so long as all the Target info and all directories are
+        added.
+        """
+        first = Resources(MockNotifier())
+        middle = Resources(MockNotifier())
+        last = Resources(MockNotifier())
+        first._add_labels('TARGET', ['K64F'])
+        first._add_labels('TARGET', ['FRDM'])
+        middle._add_labels('TARGET', ['FRDM'])
+        for name, loc in SRC_PATHS.items():
+            first.add_directory(loc, into_path=name)
+            middle.add_directory(loc, into_path=name)
+            last.add_directory(loc, into_path=name)
+        middle._add_labels('TARGET', ['K64F'])
+        last._add_labels('TARGET', ['K64F'])
+        last._add_labels('TARGET', ['FRDM'])
+        for ftype in Resources.ALL_FILE_TYPES:
+            assert(set(first.get_file_refs(ftype))
+                   == set(middle.get_file_refs(ftype)))
+            assert(set(last.get_file_refs(ftype))
+                   == set(middle.get_file_refs(ftype)))
+
+    def test_detect_duplicates(self):
+        """
+        Verify that detect_duplicates finds all of the duplicate object files
+        in the scanned tree.
+        """
+        notifier = MockNotifier()
+        first = Resources(notifier)
+        first._add_labels('TARGET', ['K64F'])
+        for name, loc in SRC_PATHS.items():
+            first.add_directory(loc, into_path=name)
+        notifier.messages = []
+        first.detect_duplicates()
+        error_messages = "\n".join(
+            m['message'] for m in notifier.messages if m['type'] == 'tool_error'
+        )
+        assert(" eggs.o " in error_messages)
+        first._add_labels('TARGET', ['FRDM'])
+        first.detect_duplicates()
+        error_messages = "\n".join(
+            m['message'] for m in notifier.messages if m['type'] == 'tool_error'
+        )
+        assert(" eggs.o " in error_messages)
+        assert(" not-main.o " in error_messages)
+        assert(" main.o " in error_messages)
+
+    def test_filter_by_all_libraries(self):
+        """
+        Assert something
+        """
+        res = Resources(MockNotifier())
+        res._add_labels('TARGET', ['K64F', 'FRDM'])
+        for name, loc in SRC_PATHS.items():
+            res.add_directory(loc, into_path=name)
+        res.filter_by_libraries(res.get_file_refs(FileType.JSON))
+        assert("main.cpp" in res.get_file_names(FileType.CPP_SRC))
+
+    def test_filter_by_bm_lib(self):
+        res = Resources(MockNotifier())
+        res._add_labels('TARGET', ['K64F', 'FRDM'])
+        for name, loc in SRC_PATHS.items():
+            res.add_directory(loc, into_path=name)
+        filter_by = [
+            ref for ref in res.get_file_refs(FileType.JSON)
+            if join("platform", "bm", "mbed_lib.json") in ref.name
+        ]
+        res.filter_by_libraries(filter_by)
+        assert("main.cpp" not in res.get_file_names(FileType.CPP_SRC))
```

Comment:
I found this confusing.
There is a main.cpp at `mbed-os.lib/TARGET_FRDM/` (which should be included) and one at `source/` (which should not be included due to the presence of `mbed_lib.json`).  But that still leaves one main.cpp, therefore I would expect this assert to fail. Or is this specifically asserting that the `main.cpp` in `source` isn't being included?

I guess my suggestion would be to rename the `main.cpp` in `mbed-os.lib/TARGET_FRDM/` for clarity.

Reply:
Yep. I'd rename that, but that's used by detect duplicates.

Reply:
AH I see, may be worth a comment here then.